src/Microsoft.Windows.ComputeVirtualization/HcsNotification.cs

```diff
@@ -0,0 +1,100 @@
+
+namespace Microsoft.Windows.ComputeVirtualization
+{
+    internal class NotificationResult
+    {
+        public int Status;
+        public string Data;
+    }
+
+    internal enum HCS_NOTIFICATIONS : uint
+    {
+        HcsNotificationInvalid = 0x00000000,
+
+        /// Notifications for HCS_SYSTEM handles
+        HcsNotificationSystemExited = 0x00000001,
+        HcsNotificationSystemCreateCompleted = 0x00000002,
+        HcsNotificationSystemStartCompleted = 0x00000003,
+
+        /// Notifications for HCS_PROCESS handles
+        HcsNotificationProcessExited = 0x00010000,
+
+        /// Common notifications
+        HcsNotificationServiceDisconnect = 0x01000000,
+
+        /// The upper 4 bits are reserved for flags. See HCS_NOTIFICATION_FLAGS
+        HcsNotificationFlagsReserved = 0xF0000000
+    }
+
+    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+    internal delegate void NotificationCallback(uint notificationType, IntPtr context, int notificationStatus, [MarshalAs(UnmanagedType.LPWStr)] string notificationData);
+
+    internal delegate int RegisterHcsNotificationCallback(IntPtr handle, NotificationCallback callback, IntPtr context, out IntPtr callbackHandle);
+
+    internal delegate int UnregisterHcsNotificationCallback(IntPtr handle);
+
+    internal class HcsNotificationWatcher : IDisposable
+    {
+        private IntPtr _h;
+        private UnregisterHcsNotificationCallback _unreg;
+        private readonly Dictionary<HCS_NOTIFICATIONS, TaskCompletionSource<NotificationResult>> _n = new Dictionary<HCS_NOTIFICATIONS, TaskCompletionSource<NotificationResult>>();
+
+        public HcsNotificationWatcher(IntPtr handle, RegisterHcsNotificationCallback register, UnregisterHcsNotificationCallback unregister, HCS_NOTIFICATIONS[] notificationList)
+        {
+            _unreg = unregister;
+
+            foreach (var notificationType in notificationList)
+            {
+                var entry = new TaskCompletionSource<NotificationResult>();
+                _n.Add(notificationType, entry);
+            }
+
+            NotificationCallback callbackFunc = (uint nType, IntPtr ctx, int nStatus, string nData) =>
+            {
+                var key = (HCS_NOTIFICATIONS)nType;
+                if (_n.ContainsKey(key))
+                {
+                    var result = new NotificationResult()
+                    {
+                        Status = nStatus,
+                        Data = nData
+                    };
+                    if (HcsException.Failed(result.Status))
+                    {
+                        _n[key].SetException(new HcsException(result.Status, result.Data));
+                    }
+                    else
+                    {
+                        _n[key].SetResult(result);                        
+                    }
+                }
+            };
+
+            HcsFunctions.ProcessHcsCall(register(handle, callbackFunc, IntPtr.Zero, out _h), null);
```

Comment:
Yeah.  There is no event to wait on that will tell us we've completed setting up event registration, so it must be synchronously completed.